tensorflow/python/keras/layers/wrappers_test.py

```diff
@@ -472,6 +472,59 @@ def test_TimeDistributed_set_static_shape(self):
     # Make sure the batch dim is not lost after array_ops.reshape.
     self.assertListEqual(outputs.shape.as_list(), [1, None, 30, 30, 16])
 
+  @keras_parameterized.run_all_keras_modes
+  def test_TimeDistributed_with_mimo(self):
+    class TestLayer(keras.layers.Layer):
+
+      def __init__(self):
+        super(TestLayer, self).__init__()
+        self.dense_1 = keras.layers.Dense(8)
+        self.dense_2 = keras.layers.Dense(16)
+
+      def build(self, input_shape):
+        super(TestLayer, self).build(input_shape)
+
+      def call(self, inputs):
+        return self.dense_1(inputs[0]), self.dense_2(inputs[1])
+
+      def compute_output_shape(self, input_shape):
+        output_shape_1 = self.dense_1.compute_output_shape(input_shape[0])
+        output_shape_2 = self.dense_2.compute_output_shape(input_shape[1])
+        return output_shape_1, output_shape_2
+
+    np.random.seed(100)
+    layer = TestLayer()
+
+    data_1 = array_ops.constant(
+        [[[[1.0], [1.0]], [[2.0], [2.0]]],
+         [[[4.0], [4.0]], [[5.0], [5.0]]],
+         [[[7.0], [7.0]], [[8.0], [8.0]]]])
+
+    data_2 = array_ops.constant(
+        [[[[1.0], [1.0]], [[2.0], [2.0]]],
+         [[[4.0], [4.0]], [[5.0], [5.0]]],
+         [[[7.0], [7.0]], [[8.0], [8.0]]]])
+
+    x1 = keras.Input(shape=(None, 2, 1), dtype='float32')
+    x2 = keras.Input(shape=(None, 2, 1), dtype='float32')
+    y1, y2 = keras.layers.TimeDistributed(layer)([x1, x2])
+    model_1 = keras.models.Model([x1, x2], [y1, y2])
+    model_1.compile(optimizer='rmsprop', loss='mse',
+                    run_eagerly=testing_utils.should_run_eagerly())
+    output_1 = model_1.predict((data_1, data_2), steps=1)
+    model_1.fit(
+      x=[np.random.random((10, 2, 2, 1)), np.random.random((10, 2, 2, 1))],
+      y=[np.random.random((10, 2, 2, 8)), np.random.random((10, 2, 2, 16))],
+      epochs=1,
+      batch_size=3)
+
+    y1 = keras.layers.Dense(8)(x1)
+    y2 = keras.layers.Dense(16)(x2)
+    model_2 = keras.models.Model([x1, x2], [y1, y2])
+    output_2 = model_2.predict((data_1, data_2), steps=1)
```

Comment:
The test case failed with following error, see https://source.cloud.google.com/results/invocations/db6264fc-58c7-4c17-aff3-f77f5eb98207/log

I1110 17:33:21.432795 140714311898880 test_util.py:2076] time(__main__.TimeDistributedTest.test_TimeDistributed_with_mimo_v2_function_use_keras_tensors): 1.8s
[  FAILED  ] TimeDistributedTest.test_TimeDistributed_with_mimo_v2_function_use_keras_tensors
[ RUN      ] TimeDistributedTest.test_timedistributed_conv2d
INFO:tensorflow:time(__main__.TimeDistributedTest.test_timedistributed_conv2d): 0.38s
I1110 17:33:21.863598 140714311898880 test_util.py:2076] time(__main__.TimeDistributedTest.test_timedistributed_conv2d): 0.38s
[       OK ] TimeDistributedTest.test_timedistributed_conv2d
======================================================================
FAIL: test_TimeDistributed_with_mimo_v2_function_use_keras_tensors (__main__.TimeDistributedTest)
test_TimeDistributed_with_mimo_v2_function_use_keras_tensors (__main__.TimeDistributedTest)
test_TimeDistributed_with_mimo_v2_function_use_keras_tensors('v2_function_use_keras_tensors')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2643, in _assertAllCloseRecursive
    a_as_ndarray = self._GetNdArray(a)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2560, in _GetNdArray
    return np.array(a)
ValueError: could not broadcast input array from shape (3,2,2,8) into shape (3,2,2)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/absl_py/absl/testing/parameterized.py", line 265, in bound_param_test
    test_method(self, *testcase_params)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/keras/keras_parameterized.py", line 417, in decorated
    _v2_function_and_kerastensors_test(f, self, *args, **kwargs)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/keras/keras_parameterized.py", line 449, in _v2_function_and_kerastensors_test
    f(test_or_class, *args, **kwargs)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/keras/layers/wrappers_test.py", line 525, in test_TimeDistributed_with_mimo
    self.assertAllClose(output_1, output_2)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1236, in decorated
    return f(*args, **kwds)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2711, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2660, in _assertAllCloseRecursive
    a_ele, b_ele, rtol=rtol, atol=atol, path=path, msg=msg)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2671, in _assertAllCloseRecursive
    (path_str, path_str, msg)))
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/layers/wrappers_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2606, in _assertArrayLikeAllClose
    a, b, rtol=rtol, atol=atol, err_msg="\n".join(msgs), equal_nan=True)
  File "/usr/local/lib/python3.6/dist-packages/numpy/testing/nose_tools/utils.py", line 1396, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/usr/local/lib/python3.6/dist-packages/numpy/testing/nose_tools/utils.py", line 779, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-06, atol=1e-06
Mismatched value: a[0] is different from b[0].
not close where = (array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2]), array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1]), array([0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1]), array([0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5,
       6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3,
       4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1,
       2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7,
       0, 1, 2, 3, 4, 5, 6, 7]))
not close lhs = [ 0.13657945 -0.33112222  0.19477272 -0.02647561  0.60698915 -0.2886145
  0.39819598  0.5779674   0.13657945 -0.33112222  0.19477272 -0.02647561
  0.60698915 -0.2886145   0.39819598  0.5779674   0.2731589  -0.66224444
  0.38954544 -0.05295122  1.2139783  -0.577229    0.79639196  1.1559348
  0.2731589  -0.66224444  0.38954544 -0.05295122  1.2139783  -0.577229
  0.79639196  1.1559348   0.5463178  -1.3244889   0.7790909  -0.10590243
  2.4279566  -1.154458    1.5927839   2.3118696   0.5463178  -1.3244889
  0.7790909  -0.10590243  2.4279566  -1.154458    1.5927839   2.3118696
  0.68289727 -1.655611    0.9738636  -0.13237804  3.0349457  -1.4430726
  1.9909799   2.889837    0.68289727 -1.655611    0.9738636  -0.13237804
  3.0349457  -1.4430726   1.9909799   2.889837    0.9560562  -2.3178556
  1.363409   -0.18532926  4.2489243  -2.0203016   2.7873719   4.0457716
  0.9560562  -2.3178556   1.363409   -0.18532926  4.2489243  -2.0203016
  2.7873719   4.0457716   1.0926356  -2.6489778   1.5581818  -0.21180487
  4.855913   -2.308916    3.1855679   4.6237392   1.0926356  -2.6489778
  1.5581818  -0.21180487  4.855913   -2.308916    3.1855679   4.6237392 ]
not close rhs = [-0.08482659  0.3962139   0.462116   -0.11807799 -0.6927713   0.21188223
 -0.71281177  0.67904234 -0.08482659  0.3962139   0.462116   -0.11807799
 -0.6927713   0.21188223 -0.71281177  0.67904234 -0.16965318  0.7924278
  0.924232   -0.23615599 -1.3855426   0.42376447 -1.4256235   1.3580847
 -0.16965318  0.7924278   0.924232   -0.23615599 -1.3855426   0.42376447
 -1.4256235   1.3580847  -0.33930635  1.5848556   1.848464   -0.47231197
 -2.7710853   0.84752893 -2.851247    2.7161694  -0.33930635  1.5848556
  1.848464   -0.47231197 -2.7710853   0.84752893 -2.851247    2.7161694
 -0.42413294  1.9810694   2.31058    -0.59038997 -3.4638567   1.0594112
 -3.5640588   3.3952117  -0.42413294  1.9810694   2.31058    -0.59038997
 -3.4638567   1.0594112  -3.5640588   3.3952117  -0.5937861   2.773497
  3.234812   -0.82654595 -4.849399    1.4831756  -4.989682    4.7532964
 -0.5937861   2.773497    3.234812   -0.82654595 -4.849399    1.4831756
 -4.989682    4.7532964  -0.6786127   3.169711    3.696928   -0.94462395
 -5.5421705   1.6950579  -5.702494    5.4323387  -0.6786127   3.169711
  3.696928   -0.94462395 -5.5421705   1.6950579  -5.702494    5.4323387 ]
not close dif = [ 0.22140604  0.7273361   0.26734328  0.09160239  1.2997605   0.50049675
  1.1110077   0.10107493  0.22140604  0.7273361   0.26734328  0.09160239
  1.2997605   0.50049675  1.1110077   0.10107493  0.4428121   1.4546722
  0.53468657  0.18320477  2.599521    1.0009935   2.2220154   0.20214987
  0.4428121   1.4546722   0.53468657  0.18320477  2.599521    1.0009935
  2.2220154   0.20214987  0.8856242   2.9093444   1.0693731   0.36640954
  5.199042    2.001987    4.444031    0.40429974  0.8856242   2.9093444
  1.0693731   0.36640954  5.199042    2.001987    4.444031    0.40429974
  1.1070302   3.6366806   1.3367164   0.45801193  6.498802    2.5024838
  5.5550385   0.50537467  1.1070302   3.6366806   1.3367164   0.45801193
  6.498802    2.5024838   5.5550385   0.50537467  1.5498424   5.0913525
  1.871403    0.6412167   9.098324    3.503477    7.777054    0.7075248
  1.5498424   5.0913525   1.871403    0.6412167   9.098324    3.503477
  7.777054    0.7075248   1.7712483   5.818689    2.1387463   0.7328191
 10.398084    4.003974    8.888062    0.8085995   1.7712483   5.818689
  2.1387463   0.7328191  10.398084    4.003974    8.888062    0.8085995 ]
not close tol = [1.0848266e-06 1.3962139e-06 1.4621160e-06 1.1180780e-06 1.6927713e-06
 1.2118822e-06 1.7128118e-06 1.6790423e-06 1.0848266e-06 1.3962139e-06
 1.4621160e-06 1.1180780e-06 1.6927713e-06 1.2118822e-06 1.7128118e-06
 1.6790423e-06 1.1696532e-06 1.7924278e-06 1.9242320e-06 1.2361560e-06
 2.3855428e-06 1.4237645e-06 2.4256235e-06 2.3580847e-06 1.1696532e-06
 1.7924278e-06 1.9242320e-06 1.2361560e-06 2.3855428e-06 1.4237645e-06
 2.4256235e-06 2.3580847e-06 1.3393063e-06 2.5848556e-06 2.8484640e-06
 1.4723120e-06 3.7710852e-06 1.8475289e-06 3.8512471e-06 3.7161694e-06
 1.3393063e-06 2.5848556e-06 2.8484640e-06 1.4723120e-06 3.7710852e-06
 1.8475289e-06 3.8512471e-06 3.7161694e-06 1.4241330e-06 2.9810694e-06
 3.3105798e-06 1.5903900e-06 4.4638568e-06 2.0594111e-06 4.5640586e-06
 4.3952118e-06 1.4241330e-06 2.9810694e-06 3.3105798e-06 1.5903900e-06
 4.4638568e-06 2.0594111e-06 4.5640586e-06 4.3952118e-06 1.5937861e-06
 3.7734972e-06 4.2348120e-06 1.8265459e-06 5.8493988e-06 2.4831756e-06
 5.9896820e-06 5.7532961e-06 1.5937861e-06 3.7734972e-06 4.2348120e-06
 1.8265459e-06 5.8493988e-06 2.4831756e-06 5.9896820e-06 5.7532961e-06
 1.6786128e-06 4.1697112e-06 4.6969280e-06 1.9446238e-06 6.5421705e-06
 2.6950579e-06 6.7024939e-06 6.4323385e-06 1.6786128e-06 4.1697112e-06
 4.6969280e-06 1.9446238e-06 6.5421705e-06 2.6950579e-06 6.7024939e-06
 6.4323385e-06]
dtype = float32, shape = (3, 2, 2, 8)
(mismatch 100.0%)
 x: array([[[[ 0.136579, -0.331122,  0.194773, -0.026476,  0.606989,
          -0.288615,  0.398196,  0.577967],
         [ 0.136579, -0.331122,  0.194773, -0.026476,  0.606989,...
 y: array([[[[-0.084827,  0.396214,  0.462116, -0.118078, -0.692771,
           0.211882, -0.712812,  0.679042],
         [-0.084827,  0.396214,  0.462116, -0.118078, -0.692771,...

----------------------------------------------------------------------
Ran 10 tests in 16.136s

FAILED (failures=1)